pkg/virt-launcher/virtwrap/sriov/hostdev.go

```diff
@@ -0,0 +1,97 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package sriov
+
+import (
+	"fmt"
+
+	"kubevirt.io/kubevirt/pkg/util"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+func CreateHostDevices(vmi *v1.VirtualMachineInstance) ([]api.HostDevice, error) {
+	var hostDevices []api.HostDevice
+
+	pciAddrPool := NewPCIAddressPool(vmi.Spec.Domain.Devices.Interfaces)
+
+	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.SRIOV == nil {
+			continue
+		}
+
+		pciAddress, err := pciAddrPool.Pop(iface.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create SRIOV hostdevice for %s: %v", iface.Name, err)
+		}
+
+		hostDevice, err := createHostDevice(iface, pciAddress)
+		if err != nil {
+			return nil, err
+		}
+		hostDevices = append(hostDevices, *hostDevice)
+		log.Log.Infof("SR-IOV PCI device created: %s", pciAddress)
+	}
+	return hostDevices, nil
+}
+
+func createHostDevice(iface v1.Interface, hostPCIAddress string) (*api.HostDevice, error) {
+	hostAddr, err := decoratePciAddressField(hostPCIAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create SRIOV device for %s, host PCI: %v", iface.Name, err)
+	}
+	hostDev := &api.HostDevice{
+		Source:  api.HostDeviceSource{Address: hostAddr},
+		Type:    "pci",
+		Managed: "no",
+	}
+
+	guestPCIAddress := iface.PciAddress
+	if guestPCIAddress != "" {
+		addr, err := decoratePciAddressField(guestPCIAddress)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create SRIOV device for %s, guest PCI: %v", iface.Name, err)
+		}
+		hostDev.Address = addr
+	}
+
+	if iface.BootOrder != nil {
+		hostDev.BootOrder = &api.BootOrder{Order: *iface.BootOrder}
+	}
+
+	return hostDev, nil
+}
+
+func decoratePciAddressField(addressField string) (*api.Address, error) {
+	dbsfFields, err := util.ParsePciAddress(addressField)
+	if err != nil {
+		return nil, err
+	}
+	decoratedAddrField := &api.Address{
+		Type:     "pci",
+		Domain:   "0x" + dbsfFields[0],
+		Bus:      "0x" + dbsfFields[1],
+		Slot:     "0x" + dbsfFields[2],
+		Function: "0x" + dbsfFields[3],
+	}
+	return decoratedAddrField, nil
+}
```

Comment:
This is problematic since we already have the exact method at converter.go
Maybe it is better to keep the HostDevice objects creating at converter.go

Reply:
The current duplication is introduced in order to keep it self-contained.
This is a good candidate to be moved to a device or hostdev package which will serve both SRIOV and the other host devices.

Keeping it in the converter does not allow me to properly test everything. There was also objection to read env variables in the converter, so this was my only way out. I could have created a hostdev package, but that is too wide to cover at the moment.

Reply:
Make sense to add a comment to point out to the duplicated code ?